File: problems/637/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<double> averageOfLevels(TreeNode* root) {
        if (not root) {
            return {};
        }
        
        vector<double> ans;
        queue<TreeNode*> q;
        
        q.push(root);
        
        while (not q.empty()) {
            int levelSize = q.size();
            double curLevelMean = 0.0;
            
            for (int i = 0; i < levelSize; i++) {
                TreeNode *node = q.front();
                q.pop();
                
                curLevelMean += node->val;
                
                if (node->left) {
                    q.push(node->left);
                }
                
                if (node->right) {
                    q.push(node->right);
                }
            }
            
            ans.push_back(curLevelMean / levelSize);
        }
        
        return ans;
    }
};

File: problems/637/solution.go
package main
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func averageOfLevels(root *TreeNode) []float64 {
    if root == nil {
        return []float64{}
    }
    
    ans, q := []float64{}, []*TreeNode{root}
    
    for len(q) > 0 {
        newQ, curAvg, n := []*TreeNode{}, 0.0, float64(len(q))
        
        for _, node := range q {
            curAvg += float64(node.Val)
            
            if node.Left != nil {
                newQ = append(newQ, node.Left)
            }
            
            if node.Right != nil {
                newQ = append(newQ, node.Right)
            }
        }
        
        ans = append(ans, curAvg / n)
        q = newQ
    }
    
    return ans
}